src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -1692,6 +1694,189 @@ case class MatrixFilterEntries(child: MatrixIR, pred: IR) extends MatrixIR {
   }
 }
 
+case class TableToMatrixTable(
+  child: TableIR,
+  rowKey: IndexedSeq[String],
+  colKey: IndexedSeq[String],
+  rowFields: IndexedSeq[String],
+  colFields: IndexedSeq[String],
+  partitionKey: IndexedSeq[String],
+  nPartitions: Option[Int] = None
+) extends MatrixIR {
+  // no fields used twice
+  private val fieldsUsed = mutable.Set.empty[String]
+  (rowKey ++ colKey ++ rowFields ++ colFields).foreach { f =>
+    assert(!fieldsUsed.contains(f))
+    fieldsUsed += f
+  }
+
+  def children: IndexedSeq[BaseIR] = FastIndexedSeq(child)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): BaseIR = {
+    val IndexedSeq(newChild) = newChildren
+    TableToMatrixTable(
+      newChild.asInstanceOf[TableIR],
+      rowKey,
+      colKey,
+      rowFields,
+      colFields,
+      partitionKey,
+      nPartitions
+    )
+  }
+
+
+  // need keys for rows and cols
+  assert(rowKey.nonEmpty)
+  assert(colKey.nonEmpty)
+
+  // check partition key is appropriate and not empty
+  assert(rowKey.startsWith(partitionKey))
+  assert(partitionKey.nonEmpty)
+
+  private val rowType = TStruct((rowKey ++ rowFields).map(f => f -> child.typ.rowType.fieldByName(f).typ): _*)
+  private val colType = TStruct((colKey ++ colFields).map(f => f -> child.typ.rowType.fieldByName(f).typ): _*)
+  private val entryFields = child.typ.rowType.fieldNames.filter(f => !fieldsUsed.contains(f))
+  private val entryType = TStruct(entryFields.map(f => f -> child.typ.rowType.fieldByName(f).typ): _*)
+
+  val typ: MatrixType = MatrixType.fromParts(
+    child.typ.globalType,
+    colKey,
+    colType,
+    partitionKey,
+    rowKey,
+    rowType,
+    entryType)
+
+  def execute(hc: HailContext): MatrixValue = {
+    val prev = child.execute(hc)
+
+    val colKeyIndices = colKey.map(child.typ.rowType.fieldIdx(_)).toArray
+    val colValueIndices = colFields.map(child.typ.rowType.fieldIdx(_)).toArray
```

Comment:
Can I assume all of this is unchanged from the previous version?

Reply:
yeah, I just moved the code.

Reply:
the only changes were renames, I believe.